File: indexer/src/main/java/com/openlattice/indexing/BackgroundIndexingService.java
/*
 * Copyright (C) 2018. OpenLattice, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * You can contact the owner of the copyright at support@openlattice.com
 *
 *
 */

package com.openlattice.indexing;

import com.google.common.collect.Maps;
import com.google.common.collect.SetMultimap;
import com.hazelcast.core.HazelcastInstance;
import com.hazelcast.core.ILock;
import com.hazelcast.core.IMap;
import com.openlattice.conductor.rpc.ConductorElasticsearchApi;
import com.openlattice.data.storage.PostgresEntityDataQueryService;
import com.openlattice.edm.EntitySet;
import com.openlattice.edm.type.EntityType;
import com.openlattice.edm.type.PropertyType;
import com.openlattice.hazelcast.HazelcastMap;
import com.openlattice.neuron.audit.AuditEntitySetUtils;
import com.openlattice.postgres.DataTables;
import com.openlattice.postgres.PostgresColumn;
import com.openlattice.postgres.ResultSetAdapters;
import com.openlattice.postgres.streams.PostgresIterable;
import com.openlattice.postgres.streams.StatementHolder;
import com.zaxxer.hikari.HikariDataSource;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Map;
import java.util.Set;
import java.util.UUID;
import java.util.stream.Collectors;
import org.apache.olingo.commons.api.edm.EdmPrimitiveTypeKind;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.scheduling.annotation.Scheduled;

public class BackgroundIndexingService {
    private static final Logger logger     = LoggerFactory.getLogger( BackgroundIndexingService.class );
    private static final int    INDEX_SIZE = 100000;
    private static final int    FETCH_SIZE = 200000;

    private final HikariDataSource               hds;
    private final PostgresEntityDataQueryService dataQueryService;
    private final ConductorElasticsearchApi      elasticsearchApi;

    private final IMap<UUID, PropertyType> propertyTypes;
    private final IMap<UUID, EntityType>   entityTypes;
    private final IMap<UUID, EntitySet>    entitySets;
    private final ILock                    lock;

    public BackgroundIndexingService(
            HikariDataSource hds,
            HazelcastInstance hazelcastInstance,
            ConductorElasticsearchApi elasticsearchApi,
            PostgresEntityDataQueryService dataQueryService ) {
        this.hds = hds;
        this.dataQueryService = dataQueryService;
        this.elasticsearchApi = elasticsearchApi;

        this.propertyTypes = hazelcastInstance.getMap( HazelcastMap.PROPERTY_TYPES.name() );
        this.entityTypes = hazelcastInstance.getMap( HazelcastMap.ENTITY_TYPES.name() );
        this.entitySets = hazelcastInstance.getMap( HazelcastMap.ENTITY_SETS.name() );

        lock = hazelcastInstance.getLock( "com.openlattice.datastore.services.BackgroundIndexingService" );
    }

    private String getDirtyEntitiesQuery( UUID entitySetId ) {
        return "SELECT * FROM \"" + DataTables.entityTableName( entitySetId ) + "\" WHERE "
                + PostgresColumn.LAST_INDEX_FIELD + " < " + PostgresColumn.LAST_WRITE_FIELD;
    }

    private PostgresIterable<UUID> getDirtyEntityKeyIds( UUID entitySetId ) {
        return new PostgresIterable<>(
                () -> {
                    Connection connection = null;
                    PreparedStatement ps;
                    ResultSet rs;
                    try {
                        connection = hds.getConnection();
                        ps = connection.prepareStatement( getDirtyEntitiesQuery( entitySetId ) );
                        ps.setFetchSize( FETCH_SIZE );
                        rs = ps.executeQuery();
                        return new StatementHolder( connection, ps, rs );
                    } catch ( SQLException e ) {
                        logger.error( "Unable to instantiate postgres iterable.", e );
                        throw new IllegalStateException( "Unable to instantiate postgres iterable.", e );
                    }
                },
                resultSet -> {
                    try {
                        return ResultSetAdapters.id( resultSet );
                    } catch ( SQLException e ) {
                        logger.error( "Unable to retrieve postgres value", e );
                        throw new IllegalStateException( "Unable to instantiate postgres iterable.", e );
                    }
                }
        );
    }

    private Map<UUID, Map<UUID, PropertyType>> getPropertyTypesByEntityTypesById() {
        Map<UUID, Map<UUID, PropertyType>> result = Maps.newHashMap();

        entityTypes.entrySet().forEach( entry -> {
            UUID entityTypeId = entry.getKey();
            Map<UUID, PropertyType> propertyTypesById = propertyTypes
                    .getAll( entityTypes.get( entityTypeId ).getProperties() ).entrySet().stream()
                    .filter( ptEntry -> !ptEntry.getValue().getDatatype().equals( EdmPrimitiveTypeKind.Binary ) )
                    .collect(
                            Collectors.toMap( ptEntry -> ptEntry.getKey(), ptEntry -> ptEntry.getValue() ) );
            result.put( entityTypeId, propertyTypesById );
        } );

        return result;
    }

    @Scheduled( fixedRate = 2000 )
    public void indexUpdatedEntitySets() {
        try {
            lock.lock();
            Map<UUID, Map<UUID, PropertyType>> propertyTypesByEntityType = getPropertyTypesByEntityTypesById();

            entitySets.values().forEach( entitySet -> {
                if ( !entitySet.getName().equals( AuditEntitySetUtils.AUDIT_ENTITY_SET_NAME ) ) {

                    UUID entitySetId = entitySet.getId();
                    Map<UUID, PropertyType> propertyTypeMap = propertyTypesByEntityType
                            .get( entitySet.getEntityTypeId() );

                    PostgresIterable<UUID> entityKeyIdsToIndex = getDirtyEntityKeyIds( entitySetId );

                    PostgresIterable.PostgresIterator<UUID> toIndexIter = entityKeyIdsToIndex.iterator();

                    while ( toIndexIter.hasNext() ) {
                        var batchToIndex = getBatch( toIndexIter );
                        Map<UUID, SetMultimap<UUID, Object>> entitiesById = dataQueryService
                                .getEntitiesById( entitySetId, propertyTypeMap, batchToIndex );

                        if ( elasticsearchApi.createBulkEntityData( entitySetId, entitiesById ) ) {
                            dataQueryService.markAsIndexed( entitySetId, batchToIndex );
                        }
                    }
                }
            } );
        } finally {
            lock.unlock();
        }
    }

    private static Set<UUID> getBatch( Iterator<UUID> entityKeyIdStream ) {
        var entityKeyIds = new HashSet<UUID>( INDEX_SIZE );

        for ( int i = 0; i < INDEX_SIZE && entityKeyIdStream.hasNext(); ++i ) {
            entityKeyIds.add( entityKeyIdStream.next() );
        }
        return entityKeyIds;
    }

}

File: indexer/src/main/kotlin/com/openlattice/indexing/BackgroundIndexingService.kt
/*
 * Copyright (C) 2018. OpenLattice, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * You can contact the owner of the copyright at support@openlattice.com
 *
 *
 */

package com.openlattice.indexing

import com.google.common.base.Stopwatch
import com.hazelcast.core.HazelcastInstance
import com.hazelcast.query.Predicates
import com.openlattice.conductor.rpc.ConductorElasticsearchApi
import com.openlattice.data.storage.IndexingMetadataManager
import com.openlattice.data.storage.MetadataOption
import com.openlattice.data.storage.PostgresEntityDataQueryService
import com.openlattice.edm.EntitySet
import com.openlattice.edm.set.EntitySetFlag
import com.openlattice.edm.type.PropertyType
import com.openlattice.hazelcast.HazelcastMap
import com.openlattice.indexer.IndexerEntitySetMetadata
import com.openlattice.indexing.configuration.IndexerConfiguration
import com.openlattice.postgres.DataTables.LAST_INDEX
import com.openlattice.postgres.DataTables.LAST_WRITE
import com.openlattice.postgres.PostgresArrays
import com.openlattice.postgres.PostgresColumn.ENTITY_SET_ID
import com.openlattice.postgres.PostgresColumn.ID
import com.openlattice.postgres.PostgresColumn.PARTITION
import com.openlattice.postgres.PostgresColumn.VERSION
import com.openlattice.postgres.PostgresTable.IDS
import com.openlattice.postgres.ResultSetAdapters
import com.openlattice.postgres.mapstores.EntitySetMapstore
import com.openlattice.postgres.streams.BasePostgresIterable
import com.openlattice.postgres.streams.PreparedStatementHolderSupplier
import com.zaxxer.hikari.HikariDataSource
import org.apache.olingo.commons.api.edm.EdmPrimitiveTypeKind
import org.slf4j.LoggerFactory
import org.springframework.scheduling.annotation.Scheduled
import java.time.Instant
import java.time.OffsetDateTime
import java.util.*
import java.util.concurrent.TimeUnit
import java.util.concurrent.locks.ReentrantLock
import java.util.stream.StreamSupport
import kotlin.streams.asSequence

const val EXPIRATION_MILLIS = 60_000L
const val INDEX_RATE = 300_000L
const val FETCH_SIZE = 128_000

/** IMPORTANT! If this number is too big, elasticsearch will explode and everything will go down. Calibrate carefully. **/
const val INDEX_SIZE = 1_000

class BackgroundIndexingService(
        hazelcastInstance: HazelcastInstance,
        private val indexerConfiguration: IndexerConfiguration,
        private val hds: HikariDataSource,
        private val dataQueryService: PostgresEntityDataQueryService,
        private val elasticsearchApi: ConductorElasticsearchApi,
        private val dataManager: IndexingMetadataManager
) {
    companion object {
        private val logger = LoggerFactory.getLogger(BackgroundIndexingService::class.java)!!
    }

    private val propertyTypes = HazelcastMap.PROPERTY_TYPES.getMap( hazelcastInstance )
    private val entityTypes = HazelcastMap.ENTITY_TYPES.getMap( hazelcastInstance )
    private val entitySets = HazelcastMap.ENTITY_SETS.getMap( hazelcastInstance )

    private val indexingLocks = HazelcastMap.INDEXING_LOCKS.getMap( hazelcastInstance )

    private val taskLock = ReentrantLock()

    @Suppress("UNUSED")
    @Scheduled(fixedRate = INDEX_RATE)
    fun indexUpdatedEntitySets() {
        logger.info("Starting background indexing task.")
        //Keep number of indexing jobs under control
        if (!taskLock.tryLock()) {
            logger.info("Not starting new indexing job as an existing one is running.")
            return
        }
        try {
            ensureAllEntityTypeIndicesExist()
            if (!indexerConfiguration.backgroundIndexingEnabled) {
                logger.info("Skipping background indexing as it is not enabled.")
                return
            }
            val w = Stopwatch.createStarted()
            //We shuffle entity sets to make sure we have a chance to work share and index everything

            val lockedEntitySets = entitySets.values(
                    Predicates.notEqual<UUID, EntitySet>(EntitySetMapstore.FLAGS_INDEX, EntitySetFlag.LINKING)
            )
                    .filter { !it.isAudit }
                    .filter { tryLockEntitySet(it.id) == null }
                    .shuffled()

            val totalIndexed = lockedEntitySets
                    .parallelStream()
                    .mapToInt { indexEntitySet(IndexerEntitySetMetadata.fromEntitySet(it)) }
                    .sum()

            lockedEntitySets.forEach { deleteIndexingLock(it.id) }

            logger.info(
                    "Completed indexing {} elements in {} ms",
                    totalIndexed,
                    w.elapsed(TimeUnit.MILLISECONDS)
            )
        } finally {
            taskLock.unlock()
        }
    }

    private fun ensureAllEntityTypeIndicesExist() {
        val existingIndices = elasticsearchApi.entityTypesWithIndices
        val missingIndices = entityTypes.keys - existingIndices
        if (missingIndices.isNotEmpty()) {
            val missingEntityTypes = entityTypes.getAll(missingIndices)
            logger.info("The following entity types were missing indices: {}", missingEntityTypes.keys)
            missingEntityTypes.values.forEach { et ->
                val missingEntityTypePropertyTypes = propertyTypes.getAll(et.properties)
                elasticsearchApi.saveEntityTypeToElasticsearch(
                        et,
                        missingEntityTypePropertyTypes.values.toList()
                )
                logger.info("Created missing index for entity type ${et.type} with id ${et.id}")
            }
        }
    }

    /**
     * Preparable sql statement to select entity key ids (with last write) of an entity set for indexing.
     * Bind order is the following:
     * 1. entity set id
     * 2. partition (array)
     * 3. partition version
     */
    private fun getEntityDataKeysQuery(reindexAll: Boolean = false, getTombstoned: Boolean = false): String {
        val dirtyIdsClause = if (!reindexAll) {
            "AND ${LAST_INDEX.name} < ${LAST_WRITE.name}"
        } else {
            ""
        }
        val versionsClause = "${VERSION.name} ${if (getTombstoned) "<=" else ">"} 0"

        return "SELECT ${ID.name}, ${LAST_WRITE.name} FROM ${IDS.name} " +
                "WHERE ${ENTITY_SET_ID.name} = ? " +
                "AND ${PARTITION.name} = ANY(?) " +
                "AND $versionsClause " +
                dirtyIdsClause
    }

    private fun getEntityDataKeys(
            entitySet: IndexerEntitySetMetadata,
            reindexAll: Boolean = false,
            getTombstoned: Boolean = false
    ): BasePostgresIterable<Pair<UUID, OffsetDateTime>> {
        return BasePostgresIterable(
                PreparedStatementHolderSupplier(hds, getEntityDataKeysQuery(reindexAll, getTombstoned), FETCH_SIZE) {
                    it.setObject(1, entitySet.id)
                    it.setArray(2, PostgresArrays.createIntArray(it.connection, *entitySet.partitions))
                }
        ) { ResultSetAdapters.id(it) to ResultSetAdapters.lastWriteTyped(it) }
    }

    private fun getPropertyTypeForEntityType(entityTypeId: UUID): Map<UUID, PropertyType> {
        return propertyTypes
                .getAll(entityTypes[entityTypeId]?.properties ?: setOf())
                .filter { it.value.datatype != EdmPrimitiveTypeKind.Binary }
    }

    /**
     * Creates missing documents for non-indexed entities in entity set and deletes documents of cleared/deleted
     * entities in entity set.
     *
     * @param entitySet The entity set that is about to get indexed.
     * @param reindexAll Indicator whether it should re-index all entities found in the entity set or just the ones,
     * which are not yet indexed.
     */
    private fun indexEntitySet(entitySet: IndexerEntitySetMetadata, reindexAll: Boolean = false): Int {

        val upsertedEntityCount = indexEntitiesInEntitySet(entitySet, reindexAll, indexTombstoned = false)
        val tombstonedEntityCount = indexEntitiesInEntitySet(entitySet, reindexAll, indexTombstoned = true)

        return upsertedEntityCount + tombstonedEntityCount
    }

    /**
     * @param entitySet The entity set that is about to get indexed.
     * @param reindexAll Indicator whether it should re-index all entities found in the entity set or just the ones,
     * which are not yet indexed.
     *
     * NOTE: If it is set to true, it also won't mark entities as indexed.
     * @param indexTombstoned Indicator whether it should un-index cleared ([VERSION] < 0) and deleted ([VERSION] = 0)
     * entities or index still active ones ([VERSION] > 0).
     */
    private fun indexEntitiesInEntitySet(
            entitySet: IndexerEntitySetMetadata,
            reindexAll: Boolean = false,
            indexTombstoned: Boolean = false
    ): Int {

        val timer = Stopwatch.createStarted()
        logger.info("starting to index entity set {}", entitySet.id)

        val propertyTypes = getPropertyTypeForEntityType(entitySet.entityTypeId)

        val entityKeyIdsWithLastWrite = getEntityDataKeys(entitySet, reindexAll, indexTombstoned)

        val indexCount = StreamSupport.stream( entityKeyIdsWithLastWrite.spliterator(), false )
                .asSequence()
                .chunked(INDEX_SIZE)
                .sumBy {
                    refreshExpiration( entitySet.id )
                    if ( indexTombstoned ) {
                        unindexEntities(entitySet, it.toMap(), !reindexAll)
                    } else {
                        indexEntities(entitySet, it.toMap(), propertyTypes, !reindexAll)
                    }
                }

        logger.info(
            "indexing entity set took {} ms - entity set {} index count {}",
            timer.elapsed(TimeUnit.MILLISECONDS),
            entitySet.id,
            indexCount
        )

        return indexCount
    }

    internal fun indexEntities(
        entitySet: IndexerEntitySetMetadata,
        batchToIndex: Map<UUID, OffsetDateTime>,
        propertyTypeMap: Map<UUID, PropertyType>,
        markAsIndexed: Boolean = true
    ): Int {

        val timer = Stopwatch.createStarted()
        logger.info("starting to index batch - entity set {} batch size {}", entitySet.id, batchToIndex.size)

        val entitiesById = dataQueryService.getEntitiesWithPropertyTypeIds(
                mapOf(entitySet.id to Optional.of(batchToIndex.keys)),
                mapOf(entitySet.id to propertyTypeMap),
                mapOf(),
                EnumSet.of(MetadataOption.LAST_WRITE)
        ).toMap()

        logger.info("getting batch entities took {} ms", timer.elapsed(TimeUnit.MILLISECONDS))

        if (entitiesById.size != batchToIndex.size) {
            logger.error(
                "expected {} but got {} entities - marking as indexed to prevent infinite loop",
                batchToIndex.size,
                entitiesById.size
            )
        }

        if (
            entitiesById.isEmpty()
            || !elasticsearchApi.createBulkEntityData(entitySet.entityTypeId, entitySet.id, entitiesById)
        ) {
            logger.error("error indexing batch - entity set {} batch size {}", entitySet.id, batchToIndex.size)
            return 0
        }

        val indexCount = if (markAsIndexed) {
            dataManager.markAsIndexed(mapOf(entitySet.id to batchToIndex))
        } else {
            batchToIndex.size
        }

        logger.info(
            "indexing batch took {} ms - entity set {} batch count {}",
            timer.elapsed(TimeUnit.MILLISECONDS),
            entitySet.id,
            indexCount
        )

        return indexCount

    }

    private fun unindexEntities(
        entitySet: IndexerEntitySetMetadata,
        batchToIndex: Map<UUID, OffsetDateTime>,
        markAsIndexed: Boolean = true
    ): Int {

        val timer = Stopwatch.createStarted()
        logger.info("starting to unindex batch - entity set {} batch size {}", entitySet.id, batchToIndex.size)

        val indexCount: Int

        if (
            batchToIndex.isNotEmpty()
            && elasticsearchApi.deleteEntityDataBulk(entitySet.entityTypeId, batchToIndex.keys)
        ) {
            indexCount = if (markAsIndexed) {
                dataManager.markAsIndexed(mapOf(entitySet.id to batchToIndex))
            } else {
                batchToIndex.size
            }

            logger.info(
                "unindexing batch took {} ms - entity set {} batch count {}",
                timer.elapsed(TimeUnit.MILLISECONDS),
                entitySet.id,
                indexCount
            )
        } else {
            indexCount = 0
            logger.error("error unindexing batch - entity set {} batch size {}", entitySet.id, batchToIndex.size)
        }

        return indexCount
    }

    private fun refreshExpiration(esId: UUID) {
        try {
            indexingLocks.lock(esId)

            tryLockEntitySet(esId)
        } finally {
            indexingLocks.unlock(esId)
        }
    }

    private fun tryLockEntitySet(esId: UUID): Long? {
        return indexingLocks.putIfAbsent(
                esId,
                Instant.now().plusMillis(EXPIRATION_MILLIS).toEpochMilli(),
                EXPIRATION_MILLIS,
                TimeUnit.MILLISECONDS
        )
    }

    private fun deleteIndexingLock(esId: UUID) {
        try {
            indexingLocks.lock(esId)
            indexingLocks.delete(esId)
        } finally {
            indexingLocks.unlock(esId)
        }
    }
}
